packages/integration-sdk-runtime/src/storage/FileSystemGraphObjectStore/FileSystemGraphObjectStore.ts

```diff
@@ -12,54 +12,90 @@ import {
+export const DEFAULT_GRAPH_OBJECT_BUFFER_THRESHOLD = 500;
 
 // it is important that this value is set to 1
 // to ensure that only one operation can be performed at a time.
 const BINARY_SEMAPHORE_CONCURRENCY = 1;
 
+export interface FileSystemGraphObjectStoreParams {
+  /**
+   * The maximum number of graph objects that this store can buffer in memory
+   * before writing to disk. Machines with more memory should consider bumping
+   * this value up.
+   *
+   * Default: 500
+   */
+  graphObjectBufferThreshold?: number;
+
+  /**
+   * Whether the files that are written to disk should be minified or not
+   */
+  prettifyFiles?: boolean;
+}
+
 export class FileSystemGraphObjectStore implements GraphObjectStore {
-  semaphore: Sema;
-  entityStorageMap: BucketMap<Entity>;
-  relationshipStorageMap: BucketMap<Relationship>;
+  private readonly semaphore: Sema;
+  private readonly localGraphObjectStore = new InMemoryGraphObjectStore();
+  private readonly graphObjectBufferThreshold: number;
+  private readonly prettifyFiles: boolean;
 
-  constructor() {
-    this.entityStorageMap = new BucketMap();
-    this.relationshipStorageMap = new BucketMap();
+  constructor(params?: FileSystemGraphObjectStoreParams) {
     this.semaphore = new Sema(BINARY_SEMAPHORE_CONCURRENCY);
+    this.graphObjectBufferThreshold =
+      params?.graphObjectBufferThreshold ||
+      DEFAULT_GRAPH_OBJECT_BUFFER_THRESHOLD;
+    this.prettifyFiles = params?.prettifyFiles || false;
   }
 
-  async addEntities(storageDirectoryPath: string, newEntities: Entity[]) {
-    this.entityStorageMap.add(storageDirectoryPath, newEntities);
+  async addEntities(
+    storageDirectoryPath: string,
+    newEntities: Entity[],
+    onEntitiesFlushed?: (entities: Entity[]) => Promise<void>,
+  ) {
+    this.localGraphObjectStore.addEntities(storageDirectoryPath, newEntities);
 
-    if (this.entityStorageMap.totalItemCount >= GRAPH_OBJECT_BUFFER_THRESHOLD) {
-      await this.flushEntitiesToDisk();
+    if (
+      this.localGraphObjectStore.getTotalEntityItemCount() >=
+      this.graphObjectBufferThreshold
+    ) {
+      await this.flushEntitiesToDisk(onEntitiesFlushed);
     }
   }
 
   async addRelationships(
     storageDirectoryPath: string,
     newRelationships: Relationship[],
+    onRelationshipsFlushed?: (relationships: Relationship[]) => Promise<void>,
   ) {
-    this.relationshipStorageMap.add(storageDirectoryPath, newRelationships);
+    this.localGraphObjectStore.addRelationships(
+      storageDirectoryPath,
+      newRelationships,
+    );
 
     if (
-      this.relationshipStorageMap.totalItemCount >=
-      GRAPH_OBJECT_BUFFER_THRESHOLD
+      this.localGraphObjectStore.getTotalRelationshipItemCount() >=
+      this.graphObjectBufferThreshold
     ) {
-      await this.flushRelationshipsToDisk();
+      await this.flushRelationshipsToDisk(onRelationshipsFlushed);
     }
   }
 
   async getEntity({ _key, _type }: GraphObjectLookupKey): Promise<Entity> {
```

Comment:
Thought: what information is typically gotten from previous entities? My thought would be that it would usually only be foreign keys, which is likely just an id. Is it worth considering when we flush `InMemoryGraphObjectStore`, we maintain a map similar to the implementation of `DuplicateKeyTracker`. 
Maybe we could get lucky and reduce roundtrips to the disk a good amount of the time. ¯\_(ツ)_/¯

Reply:
There are definitely lots of additional caching improvements that we can make!